crates/base/src/rt_worker/supervisor/strategy_per_worker.rs

```diff
@@ -0,0 +1,144 @@
+use std::{sync::atomic::Ordering, time::Duration};
+
+use event_worker::events::ShutdownReason;
+use log::error;
+use sb_workers::context::{Timing, TimingStatus, UserWorkerMsgs};
+
+use super::{handle_interrupt, Arguments, IsolateInterruptData};
+
+pub async fn supervise(args: Arguments) -> ShutdownReason {
+    let Arguments {
+        key,
+        runtime_opts,
+        timing,
+        mut cpu_alarms_rx,
+        mut memory_limit_rx,
+        pool_msg_tx,
+        isolate_memory_usage_tx,
+        thread_safe_handle,
+        ..
+    } = args;
+
+    let Timing {
+        status: TimingStatus { demand, is_retired },
+        req: (_, mut req_end_rx),
+    } = timing.unwrap_or_default();
+
+    let is_retired = is_retired.unwrap();
+
+    let mut cpu_time_soft_limit_reached = false;
+    let mut wall_clock_alerts = 0;
+    let mut req_ack_count = 0usize;
+
+    // reduce 100ms from wall clock duration, so the interrupt can be handled before
+    // isolate is dropped
+    let wall_clock_duration =
+        Duration::from_millis(runtime_opts.worker_timeout_ms) - Duration::from_millis(100);
+
+    // Split wall clock duration into 2 intervals.
+    // At the first interval, we will send a msg to retire the worker.
+    let wall_clock_duration_alert = tokio::time::interval(
+        wall_clock_duration
+            .checked_div(2)
+            .unwrap_or(Duration::from_millis(0)),
+    );
+
+    tokio::pin!(wall_clock_duration_alert);
+
+    loop {
+        tokio::select! {
+            Some(_) = cpu_alarms_rx.recv() => {
+                if !cpu_time_soft_limit_reached {
+                    // retire worker
+                    is_retired.raise();
+                    error!("CPU time soft limit reached. isolate: {:?}", key);
+                    cpu_time_soft_limit_reached = true;
+
+                    if req_ack_count == demand.load(Ordering::Acquire) {
+                        let interrupt_data = IsolateInterruptData {
```

Comment:
This struct seems like it can go outside, it's also re-used during the `else`

Reply:
Yes, it's probably reusable.
In the case of `per_worker`, It just nothing but moved into the separated module, so I'm not touched aggressively before.

I'll fix it